File: algorithm/src/P87377.kt
import kotlin.math.max
import kotlin.math.min

class P87377 {
    fun solution(line: Array<IntArray>): Array<String> {
        val starArr = mutableListOf<Pair<Int, Int>>()

        var minX = Int.MAX_VALUE
        var minY = Int.MAX_VALUE
        var maxX = Int.MIN_VALUE
        var maxY = Int.MIN_VALUE

        for (i in 0 until line.size) {
            val (a, b, e) = line[i]
            for (j in i + 1 until line.size) {
                val (c, d, f) = line[j]

                // Long으로 변환해서 연산
                val adbc = a.toLong() * d - b.toLong() * c
                if (adbc == 0L) continue // 평행한 경우

                val bfed = b.toLong() * f - e.toLong() * d
                val ecaf = e.toLong() * c - a.toLong() * f

                val x = bfed.toDouble() / adbc.toDouble()
                val y = ecaf.toDouble() / adbc.toDouble()

                val intX = x.toInt()
                val intY = y.toInt()
                if (x == intX.toDouble() && y == intY.toDouble()) {
                    starArr.add(Pair(intX, intY))

                    minX = min(minX, intX)
                    minY = min(minY, intY)
                    maxX = max(maxX, intX)
                    maxY = max(maxY, intY)
                }
            }
        }

        val xLen = maxX - minX + 1
        val yLen = maxY - minY + 1
        val coord = Array(yLen) { CharArray(xLen) { '.' } }

        for ((starX, starY) in starArr) {
            val nx = starX - minX
            val ny = starY - minY
            coord[ny][nx] = '*'
        }

        val answer = mutableListOf<String>()
        for (result in coord) {
            answer.add(result.joinToString(""))
        }

        return answer.reversed().toTypedArray()
    }
}

fun main() {
    val pr = P87377()
    pr.solution(
        arrayOf(
            intArrayOf(2, -1, 4),
            intArrayOf(-2, -1, 4),
            intArrayOf(0, -1, 1),
            intArrayOf(5, -8, -12),
            intArrayOf(5, 8, 12)
        )
    )
}

File: algorithm/python/P87377.py
def solution(line):
    pos, answer = [], []

    n = len(line)

    min_x = min_y = int(1e15)
    max_x = max_y = int(-1e15)

    for i in range(n):
        a, b, e = line[i]
        for j in range(i + 1, n):
            c, d, f = line[j]

            if a * d == b * c: continue

            x = (b * f - e * d) / (a * d - b * c)
            y = (e * c - a * f) / (a * d - b * c)

            if (x == int(x) and y == int(y)):
                x = int(x)
                y = int(y)
                pos.append([x, y])

                if min_x > x: min_x = x
                if min_y > y: min_y = y
                if max_x < x: max_x = x
                if max_y < y: max_y = y


    x_len = max_x - min_x + 1
    y_len = max_y - min_y + 1

    coord = [['.'] * x_len for _ in range(y_len)]

    for star_x, star_y in pos:
        nx = star_x + abs(min_x) if max_x < 0 else star_x - min_x
        ny = star_y + abs(min_y) if min_y < 0 else star_y - min_y
        coord[ny][nx] = '*'

    for result in coord: answer.append(''.join(result))

    return answer[::-1]